### PR TITLE
Clean up SqlQueryManager

### PR DESCRIPTION
**Pull request by dain**
* Fix warnings
* Extract out complex book keeping like query timeout tracking
* Simplify session creation code
* Prepare manager for split into dispatcher and coordinator

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 1]
```diff
@@ -41,6 +41,7 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.TimeZone;
+import java.util.stream.Collectors;
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: `Application of defaults during session creation conflicts with resource
group selection and query parsing. Separating out the application of
default session properties allows us to refactor query creation in
SqlQueryManager.`

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 2]
```diff
@@ -344,6 +345,59 @@ public Session beginTransactionId(TransactionId transactionId, TransactionManage
...
 
+    public Session applyDefaultProperties(Map<String, String> systemPropertyDefaults, Map<String, Map<String, String>> catalogPropertyDefaults)
+    {
+        requireNonNull(systemPropertyDefaults, "systemPropertyDefaults is null");
+        requireNonNull(catalogPropertyDefaults, "catalogPropertyDefaults is null");
+
+        // to remove this check properties must be authenticated and validated as in beginTransactionId
+        checkState(
+                !this.transactionId.isPresent() && this.connectorProperties.isEmpty(),
+                "Session properties can not be overridden once a transaction is active");
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: "cannot"

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 3]
```diff
@@ -344,6 +345,59 @@ public Session beginTransactionId(TransactionId transactionId, TransactionManage
...
+
+        Map<String, String> systemProperties = new HashMap<>();
+        for (Entry<String, String> entry : systemPropertyDefaults.entrySet()) {
+            systemProperties.put(entry.getKey(), entry.getValue());
+        }
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: Why wouldn't you just do:
```
systemProperties.putAll(systemPropertyDefaults);
systemProperties.putAll(this.systemProperties);
```

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 4]
```diff
@@ -344,6 +345,59 @@ public Session beginTransactionId(TransactionId transactionId, TransactionManage
                 preparedStatements);
     }
 
+    public Session applyDefaultProperties(Map<String, String> systemPropertyDefaults, Map<String, Map<String, String>> catalogPropertyDefaults)
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: applyDefaultProperties to me indicates that you are going to modify the instance. Perhaps something like `withDefaultProperties` is more indicative that it returns a new object.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 5]
```diff
@@ -425,7 +429,8 @@ public QueryId createQueryId()
                     sessionContext.getResourceEstimates(),
                     queryType));
 
-            session = sessionSupplier.createSession(queryId, sessionContext, queryType, selectionContext.getResourceGroupId());
+            session = sessionSupplier.createSession(queryId, sessionContext);
+            session = sessionPropertyDefaults.applyDefaultProperties(session, queryType, selectionContext.getResourceGroupId());
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: I would also rename the method in SessionPropertyDefaults

**Comment by dain** [comment]
The changes to auto commit handling depend on #11511.   The `TestTpchDistributedStats` ends up using `com.facebook.presto.tests.DistributedQueryRunner#createPlan` which cancel's the query, triggering the auto commit cleanup, so later when `com.facebook.presto.tests.statistics.MetricComparator#getEstimatedValuesInternal` tries to calculate stats and access the metadata, and error is thrown because the transaction is gone.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+     */
+    public void expireQuery(QueryId queryId)
+    {
+        tryGetQuery(queryId)
+                .ifPresent(expirationQueue::add);
+    }
+
+    @Managed(description = "Query garbage collector executor")
+    @Nested
+    public ThreadPoolExecutorMBean getManagementExecutor()
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: How would this appear in debugging/ops tools? What would the object name be?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+    /**
+     * Enforce query max runtime/execution time limits
+     */
+    private void enforceTimeLimits()
+    {
+        for (T query : queries.values()) {
+            if (query.isDone()) {
+                continue;
+            }
+            Duration queryMaxRunTime = SystemSessionProperties.getQueryMaxRunTime(query.getSession());
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: I know it looked like this previously, but shouldn't these be static imported?

File: presto-main/src/main/java/com/facebook/presto/execution/FailedQueryExecution.java [hunk 8]
```diff
@@ -45,16 +48,37 @@
...
+    public static FailedQueryExecution createFailedQueryExecution(
+            String query,
+            Session session,
+            URI self,
+            TransactionManager transactionManager,
+            Executor executor,
+            Metadata metadata,
+            Throwable cause)
+    {
+        QUERY_STATE_LOG.debug(cause, "Query %s failed", session.getQueryId());
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/FailedQueryExecution.java]: I've never seen a public logger shared between classes. Not sure why we went in this direction.

File: presto-main/src/main/java/com/facebook/presto/execution/FailedQueryExecution.java [hunk 9]
```diff
@@ -222,7 +246,7 @@ public void pruneInfo()
     @Override
     public Optional<ResourceGroupId> getResourceGroup()
     {
-        return resourceGroup;
+        return Optional.empty();
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/FailedQueryExecution.java]: Don't all failed queries become a FailedQueryExecution (including ones with resource groups)?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 10]
```diff
@@ -755,8 +755,9 @@ public boolean transitionToFinishing()
             return false;
         }
 
-        if (autoCommit) {
-            ListenableFuture<?> commitFuture = transactionManager.asyncCommit(session.getTransactionId().get());
+        Optional<TransactionId> transactionId = session.getTransactionId();
+        if (transactionId.isPresent() && transactionManager.transactionExists(transactionId.get()) && transactionManager.isAutoCommit(transactionId.get())) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: I thought there might be a race between the two calls, but it seems the only way to remove the transaction is to call `asyncAbort` which is guarded by the state machine.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 11]
```diff
@@ -212,25 +212,15 @@ static QueryStateMachine beginWithTicker(
             Ticker ticker,
             Metadata metadata)
     {
-        session.getTransactionId().ifPresent(transactionControl ? transactionManager::trySetActive : transactionManager::checkAndSetActive);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: I think we can get rid of `transactionControl` now, here and all the way up, including in `DataDefinitionTask`.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 10 above)
Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: the extra call is required for `ROLLBACK` which immediately removes the transaction from the manager, so without the exists check the call to `isAutoCommit` fails.  I looked as having an explicit check for `ROLLBACK` here, but that would require breaking to many abstractions.  I also looked at making `isAutoCommit`, return Optional, or just false, for unknown queries, but that makes is difficult to work with in the normal case.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 6 above)
Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: Actually, this is totally wrong.  QueryTracker is not exported to JMX, so it won't show up at all.  I'm going to change query tracker to take the executor in the constructor, and will use the existing executor in the `SqlQueryManager`, so there will be no change to the existing JMX exports.

File: presto-main/src/main/java/com/facebook/presto/execution/FailedQueryExecution.java (hunk 9 above)
Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/execution/FailedQueryExecution.java]: It makes the code a bit messier. How important is it to have this?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 11 above)
Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: @electrum `transactionControl` is used on the next line of this code: `if (!session.getTransactionId().isPresent() && !transactionControl) {`

**Comment by dain** [comment]
@raghavsethi changes applied. Please review again.

**Comment by raghavsethi** [comment]
Looks good to merge after rebase + tests.

**Pull request closed without merging**